README.md
# Disaster_Recognition_App_CNN

In this project I would like to present the application based on the convolutional neural network, able to recognition a natural disaster like: cyclone, flood, earthquake and wildfire.

The data for training, validation and testing, that contains almost 4,500 images, has been downloaded from the below:
https://drive.google.com/file/d/1NvTyhUsrFbL91E10EPm38IjoCg6E2c6q/view

The project contains 4 python files, used for:
- paths_creating.py - creating all neccessary paths on a drive
- images_copying.py - splitting images into the train/validation/test subfolders
- model.py - the main model (based on keras cnn)
- model_interface.py - the interface written using TkInter to operate the model on a new images.

Moreover, in the repository one can find the trained model  - best_model.h5 (together with the history file, which can be used to run the model_interface.py without any further trainin), screens that can be found below and some sample images that the model recognizes.


![App screen](https://github.com/Yelon-ml/Disaster_Recognition_App_CNN/blob/master/app%20screen/app1.PNG)
![App screen](https://github.com/Yelon-ml/Disaster_Recognition_App_CNN/blob/master/app%20screen/app2.PNG)


history
"{'val_loss': [0.7147377133369446, 0.7630096673965454, 0.7740507125854492, 0.45617204904556274, 0.6830905079841614, 0.22752255201339722, 0.608212411403656, 0.4317352771759033, 0.449190616607666, 0.2717137634754181, 0.3070180118083954, 0.4687957465648651, 0.05012744292616844, 0.3715546131134033, 0.21356509625911713, 0.5316712856292725, 0.37585321068763733], 'val_acc': [0.6707473993301392, 0.7389323115348816, 0.7252604365348816, 0.794921875, 0.7233073115348816, 0.830078125, 0.8298969268798828, 0.8600260615348816, 0.8684895634651184, 0.841796875, 0.8834635615348816, 0.8359375, 0.8450520634651184, 0.8453608155250549, 0.8463541865348816, 0.84765625, 0.849609375], 'loss': [1.0873644808928171, 0.8150570456186931, 0.7247294427951178, 0.6425425658623377, 0.5985293218493462, 0.5632734107971191, 0.5344125850995382, 0.48803281972805657, 0.45872750798861184, 0.44465626309315365, 0.4217903054753939, 0.4188076798121134, 0.39132188240687055, 0.38745866785446803, 0.3702289247512817, 0.34991430848836896, 0.3468682083487511], 'acc': [0.5104167, 0.669375, 0.71166664, 0.74875, 0.77375, 0.79270834, 0.79895836, 0.82125, 0.82916665, 0.83479166, 0.8379167, 0.85020834, 0.8620833, 0.865, 0.86270833, 0.87125, 0.87333333]}"

model.py
import os
import sys
import importlib
from paths_creating import *
import numpy as np
from datetime import datetime
import json

import keras
from keras import layers
from keras import models
from keras import optimizers
from keras.models import load_model
from keras.preprocessing import image
from keras.preprocessing.image import ImageDataGenerator
from keras import callbacks
from keras.callbacks import EarlyStopping
from keras.callbacks import ModelCheckpoint

import matplotlib.pyplot as plt

print(len(os.listdir(train_dir_c)), len(os.listdir(train_dir_e)), len(os.listdir(train_dir_f)), len(os.listdir(train_dir_w)))
print(len(os.listdir(validation_dir_c)), len(os.listdir(validation_dir_e)), len(os.listdir(validation_dir_f)), len(os.listdir(validation_dir_w)))
print(len(os.listdir(test_dir_c)), len(os.listdir(test_dir_e)), len(os.listdir(test_dir_f)), len(os.listdir(test_dir_w)))

class Model:

    def __init__(self):
        self.model = None
        self.model_path = 'best_model.h5'
        self.model_history_path = 'history'
        self.model_exists = os.path.exists(self.model_path)
        self.history_exists = os.path.exists(self.model_history_path)

    def create_model(self):
        if self.model_exists:
            self.model = load_model(self.model_path)
        else:
            self.model = models.Sequential()
            self.model.add(layers.Conv2D(32, (3, 3), activation = 'relu', input_shape= (180, 180, 3)))
            self.model.add(layers.MaxPooling2D((2, 2)))
            self.model.add(layers.Conv2D(64, (3, 3), activation = 'relu'))
            self.model.add(layers.MaxPooling2D((2, 2)))
            self.model.add(layers.Conv2D(128, (3, 3), activation= 'relu'))
            self.model.add(layers.MaxPooling2D((2, 2)))
            self.model.add(layers.Conv2D(128, (3, 3), activation= 'relu'))
            self.model.add(layers.MaxPooling2D((2, 2)))
            self.model.add(layers.Conv2D(256, (3, 3), activation = 'relu'))
            self.model.add(layers.MaxPooling2D((2, 2)))
            self.model.add(layers.Flatten())
            self.model.add(layers.Dropout(0.3))
            self.model.add(layers.Dense(512, activation = 'relu'))
            self.model.add(layers.Dense(4, activation = 'softmax'))

    def summary(self):
        return self.model.summary()

    def compile(self):
        if not self.model_exists:
            self.model.compile(optimizer=optimizers.RMSprop(lr=1e-4), loss='categorical_crossentropy', metrics=['acc'])
            print("Model compiled.")
        else:
            print("\n\nModel found, no compilation needed.\n\n")

    def data_generator(self):
        self.train_datagen = ImageDataGenerator(rescale= 1./255,
                                    rotation_range= 40,
                                    width_shift_range= .2,
                                    height_shift_range= .2,
                                    shear_range= .2,
                                    zoom_range= .2,
                                    horizontal_flip= True)

        self.test_datagen = ImageDataGenerator(rescale= 1./255)

        self.train_generator = self.train_datagen.flow_from_directory(train_dir,
                                                    target_size= (180, 180),
                                                    batch_size=32,
                                                    class_mode='categorical')

        self.validation_generator = self.test_datagen.flow_from_directory(validation_dir,
                                                    target_size=(180, 180),
                                                    batch_size=32,
                                                    class_mode='categorical')



    def train(self):

        self.logdir = "DisasterModel\\Logs" + datetime.now().strftime("%Y%m%d-%H%M%S")
        self.tensorboard_callback = keras.callbacks.TensorBoard(log_dir = self.logdir, histogram_freq=0, write_graph=True, write_images=True)
        self.es = EarlyStopping(monitor='val_acc', mode='max', verbose=1, patience=6)
        self.mc = ModelCheckpoint('best_model.h5', monitor='val_acc', mode='max', verbose=1, save_best_only=True)

        self.history = None

        if self.history_exists:
            with open(self.model_history_path, 'r') as json_file:
                self.history = json.load(json_file)
                print("\n\nThe history file has been found, no training needed.\n")
        else:
            self.history = self.model.fit_generator(self.train_generator,
                                                    steps_per_epoch=150,
                                                    epochs=100,
                                                    validation_data=self.validation_generator,
                                                    validation_steps=50,
                                                    callbacks=[self.es, self.mc, self.tensorboard_callback])
            print("\n\nModel has been trained.\n\n")


if __name__ == "__main__":
    disaster_model = Model()
    disaster_model.create_model()
    disaster_model.summary()
    disaster_model.compile()
    disaster_model.data_generator()
    disaster_model.train()


evaluate_model.py
from paths_creating import *
import keras
from keras import models
from keras.models import load_model
from keras.preprocessing.image import ImageDataGenerator

class Evaluation:

    def __init__(self):

        self.best_model = 'best_model.h5'
        self.model = load_model(self.best_model)

        self.test_datagen = ImageDataGenerator(rescale=1./255)
        self.test_generator = test_datagen.flow_from_directory(test_dir, target_size=(180, 180), batch_size=1)

        print("\n\nLength of the test batch: {}.".format(len(self.test_generator))) # [] batch_nr, [] data/labels, [] img_nr, [][] dims, [] channels

        self.loss_test, self.acc_test = self.model.evaluate(self.test_generator)

        print("\n\nThe test accuracy is equal to {},\nwhile the loss on the test batch is equal to {}.".format(self.acc_test, self.loss_test))

if __name__ == "__main__":
    eval = Evaluation()


paths_creating.py
import os

model_dir = r'C:\DisasterModel'

cyclone_dir = r'C:\DisasterModel\Cyclone_Wildfire_Flood_Earthquake_Database\cyclone'
earthquake_dir = r'C:\DisasterMode\Cyclone_Wildfire_Flood_Earthquake_Database\earthquake'
flood_dir = r'C:\DisasterModel\Cyclone_Wildfire_Flood_Earthquake_Database\flood'
wildfire_dir = r'C:\DisasterModel\Cyclone_Wildfire_Flood_Earthquake_Database\wildfire'

train_dir = os.path.join(model_dir, 'train')
validation_dir = os.path.join(model_dir, 'validation')
test_dir = os.path.join(model_dir, 'test')

train_dir_c = os.path.join(train_dir, 'cyclone')
train_dir_e = os.path.join(train_dir, 'earthquake')
train_dir_f = os.path.join(train_dir, 'flood')
train_dir_w = os.path.join(train_dir, 'wildfire')

validation_dir_c = os.path.join(validation_dir, 'cyclone')
validation_dir_e = os.path.join(validation_dir, 'earthquake')
validation_dir_f = os.path.join(validation_dir, 'flood')
validation_dir_w = os.path.join(validation_dir, 'wildfire')

test_dir_c = os.path.join(test_dir, 'cyclone')
test_dir_e = os.path.join(test_dir, 'earthquake')
test_dir_f = os.path.join(test_dir, 'flood')
test_dir_w = os.path.join(test_dir, 'wildfire')

create_dirs = False  #False if path already exist

if create_dirs:
    os.mkdir(train_dir)
    os.mkdir(validation_dir)
    os.mkdir(test_dir)
    os.mkdir(train_dir_c)
    os.mkdir(train_dir_e)
    os.mkdir(train_dir_f)
    os.mkdir(train_dir_w)
    os.mkdir(validation_dir_c)
    os.mkdir(validation_dir_e)
    os.mkdir(validation_dir_f)
    os.mkdir(validation_dir_w)
    os.mkdir(test_dir_c)
    os.mkdir(test_dir_e)
    os.mkdir(test_dir_f)
    os.mkdir(test_dir_w)


model_interface.py
from tkinter import *
from PIL import ImageTk, Image
from tkinter import filedialog

import keras
from keras import models
from keras.models import Model
from keras.models import load_model

import os
import time
import numpy as np
import matplotlib.pyplot as plt


class MyApp(Frame):

    def __init__(self, master, *args, **kwargs):

        self.root = master
        self.root.geometry("250x600")
        self.root.configure(background="#6cb6eb")

        ### BUTTONS ###

        self.load_image_but = Button(self.root, text='Load Image', height=3, width=28, command=self.load_image)
        self.load_image_but.grid(row = 1, column = 1)
        self.load_model_but = Button(self.root, text='Load Model', height=3, width=28, command=self.load_model_)
        self.load_model_but.grid(row = 2, column=1)
        self.prediction_but = Button(self.root, text="Prediction", height=3, width=28, command=self.prediction, state="disabled")
        self.prediction_but.grid(row=4, column=1)
        self.quit_but = Button(self.root, text="Quit", height=3, width=28, command=quit)
        self.quit_but.grid(row=10, column=1)

        self.layer1_but = Button(self.root, text='Filters in layer 1', height=3, width=28, command=lambda: self.display_activation(self.activations, 0), state="disabled")
        self.layer1_but.grid(row=6, column=1)
        self.layer2_but = Button(self.root, text='Filters in layer 2', height=3, width=28, command=lambda: self.display_activation(self.activations, 1), state="disabled")
        self.layer2_but.grid(row=7, column=1)
        self.layer3_but = Button(self.root, text='Filters in layer 3', height=3, width=28, command=lambda: self.display_activation(self.activations, 2), state="disabled")
        self.layer3_but.grid(row=8, column=1)
        self.layer4_but = Button(self.root, text='Filters in layer 4', height=3, width=28, command=lambda: self.display_activation(self.activations, 3), state="disabled")
        self.layer4_but.grid(row=9, column=1)

        self.lef_but = Button(self.root, text="<", height=3, width=8, state='disable', command=self.move_in_the_left)
        self.lef_but.grid(row=1, column=5)
        self.right_but = Button(self.root, text=">", height=3, width=8, command=self.move_in_the_right)
        self.right_but.grid(row=1, column=7)

        ### LABELS ###

        self.empty_lab0 = Label(self.root, width=2, background='#6cb6eb')
        self.empty_lab0.grid(row=0, column=0)
        self.pred_lab = Label(self.root, text="", height=2, background='#6cb6eb')
        self.pred_lab.grid(row=5, column=1)
        self.empty_lab2 = Label(self.root, background='#6cb6eb', height=2, text="")
        self.empty_lab2.grid(row=3, column=1)
        self.empty_lab3 = Label(self.root, width=3, background='#6cb6eb')
        self.empty_lab3.grid(row=4, column=2)
        self.title_lab = Label(self.root, width=66, background='#6cb6eb', text="")
        self.title_lab.grid(row=1, column=3)
        self.empty_lab4 = Label(self.root, width=2, background='#6cb6eb')
        self.empty_lab4.grid(row=11, column=0)
        self.empty_lab5 = Label(self.root, width=2, background='#6cb6eb')
        self.empty_lab5.grid(row=12, column=0)
        self.empty_lab6 = Label(self.root, width=2, background='#6cb6eb')
        self.empty_lab6.grid(row=1, column=4)
        self.path_lab = Label(self.root, background='#6cb6eb', text="")
        self.path_lab.grid(row=11, column=1, columnspan=3, rowspan=2)
        self.lab_between_lr = Label(self.root, width=44 , background='#6cb6eb')
        self.lab_between_lr.grid(row=1, column=6)
        self.filter_lab = Label(self.root, background='#6cb6eb', text="")
        self.filter_lab.grid(row=1, column=6)


        ### FUNCTIONALITY ###

    def load_image(self):
        self.x = filedialog.askopenfilename(title='Select the image')
        img = Image.open(self.x)
        img = img.resize((600, 450), Image.ANTIALIAS)
        img = ImageTk.PhotoImage(img)
        panel = Label(root, image = img)
        panel.image = img
        panel.grid(row = 2, column=3, rowspan=9, columnspan=1)
        img_title = "The " + str(os.path.basename(self.x)) + " image has been loaded."
        img_path = "The image has been loaded from the: " + str(os.path.dirname(self.x)) + "."
        self.title_lab.config(text=img_title, font=("Arial, 11"))
        self.path_lab.config(text=img_path)
        self.load_image_but.config(text="Change Image")
        self.root.geometry("875x600")
        self.layer1_but.config(state='disable')
        self.layer2_but.config(state='disable')
        self.layer3_but.config(state='disable')
        self.layer4_but.config(state='disable')

    def load_model_(self):
        start_loading = time.time()
        best_model = 'best_model.h5'
        self.model = load_model(best_model)
        loading_period = str(np.round(time.time() - start_loading, 3))
        loaded = "Model loaded in " + loading_period + " sec."
        self.empty_lab2.config(text=loaded)
        self.prediction_but.config(state='normal')

    def prediction(self):
        self.label_names = {0: 'cyclone', 1: 'earthquake', 2: 'flood', 3: 'wildfire'}
        image = keras.preprocessing.image.load_img(self.x, color_mode="rgb", target_size=(180, 180))
        input_arr = keras.preprocessing.image.img_to_array(image)
        input_arr = np.array([input_arr])
        pred = self.model.predict(input_arr)
        pred_title = self.label_names[np.argmax(pred)]
        pred_title = "The image has been classified as\n" + str(pred_title) + "."
        self.pred_lab.config(text=pred_title)
        layer_outputs = [layer.output for layer in self.model.layers]
        activation_model = Model(inputs=self.model.input, outputs=layer_outputs)
        self.activations = activation_model.predict(input_arr)
        self.layer1_but.config(state='normal')
        self.layer2_but.config(state='normal')
        self.layer3_but.config(state='normal')
        self.layer4_but.config(state='normal')

    def display_activation(self, activations, act_index):
        self.act_index = act_index
        self.activation = activations[self.act_index]
        self.activation_index=0
        fig = plt.figure(figsize=(4.5, 4.5))
        fig.patch.set_facecolor("#6cb6eb")
        plt.subplots_adjust(left=0, bottom=0, right=1, top=1, wspace=0, hspace=0)
        plt.imshow(self.activation[0, :, :, self.activation_index], cmap='gray', interpolation='nearest')
        plt.axis('off')
        fig.canvas.draw()
        data = np.fromstring(fig.canvas.tostring_rgb(), dtype=np.uint8, sep='')
        data = data.reshape(fig.canvas.get_width_height()[::-1] + (3,))
        img = ImageTk.PhotoImage(image=Image.fromarray(data))
        panel = Label(root, image = img)
        panel.image = img
        panel.grid(row = 2, column=5, rowspan=9, columnspan=3)
        filter_title = "Filter {} of {} in layer {}".format(self.activation_index + 1, self.activation[0, :, :].shape[2], self.act_index + 1)
        self.filter_lab.config(text=filter_title, font=("Arial, 10"))
        self.root.geometry("1355x600")
        if self.activation_index == 0:
            self.lef_but.config(state='disabled')

    def move_in_the_right(self):
        self.activation_index+=1
        fig = plt.figure(figsize=(4.5, 4.5))
        fig.patch.set_facecolor("#6cb6eb")
        plt.subplots_adjust(left=0, bottom=0, right=1, top=1, wspace=0, hspace=0)
        plt.imshow(self.activation[0, :, :, self.activation_index], cmap='gray')
        plt.axis('off')
        fig.canvas.draw()
        data = np.fromstring(fig.canvas.tostring_rgb(), dtype=np.uint8, sep='')
        data = data.reshape(fig.canvas.get_width_height()[::-1] + (3,))
        img = ImageTk.PhotoImage(image=Image.fromarray(data))
        panel = Label(root, image = img)
        panel.image = img
        panel.grid(row = 2, column=5, rowspan=9, columnspan=3)
        filter_title = "Filter {} of {} in layer {}".format(self.activation_index + 1, self.activation[0, :, :].shape[2], self.act_index + 1)
        self.filter_lab.config(text=filter_title)
        self.lef_but.config(state='normal')
        if self.activation_index == int(self.activation[0, :, :].shape[2] - 1):
            self.right_but.config(state='disabled')

    def move_in_the_left(self):
        self.activation_index-=1
        fig = plt.figure(figsize=(4.5, 4.5))
        fig.patch.set_facecolor("#6cb6eb")
        plt.subplots_adjust(left=0, bottom=0, right=1, top=1, wspace=0, hspace=0)
        plt.imshow(self.activation[0, :, :, self.activation_index], cmap='gray')
        plt.axis('off')
        fig.canvas.draw()
        data = np.fromstring(fig.canvas.tostring_rgb(), dtype=np.uint8, sep='')
        data = data.reshape(fig.canvas.get_width_height()[::-1] + (3,))
        img = ImageTk.PhotoImage(image=Image.fromarray(data))
        panel = Label(root, image = img)
        panel.image = img
        panel.grid(row = 2, column=5, rowspan=9, columnspan=3)
        filter_title = "Filter {} of {} in layer {}".format(self.activation_index + 1, self.activation[0, :, :].shape[2], self.act_index + 1)
        self.filter_lab.config(text=filter_title)
        self.right_but.config(state='normal')
        if self.activation_index == 0:
            self.lef_but.config(state='disabled')


if __name__ == "__main__":
    root = Tk()
    App = MyApp(root)
    root.title("Disaster Interface")
    root.mainloop()


images_copying.py
import os
import shutil
import paths_creating

train_dir = os.path.join(model_dir, 'train')
validation_dir = os.path.join(model_dir, 'validation')
test_dir = os.path.join(model_dir, 'test')

train_dir_c = os.path.join(train_dir, 'cyclone')
train_dir_e = os.path.join(train_dir, 'earthquake')
train_dir_f = os.path.join(train_dir, 'flood')
train_dir_w = os.path.join(train_dir, 'wildfire')

validation_dir_c = os.path.join(validation_dir, 'cyclone')
validation_dir_e = os.path.join(validation_dir, 'earthquake')
validation_dir_f = os.path.join(validation_dir, 'flood')
validation_dir_w = os.path.join(validation_dir, 'wildfire')

test_dir_c = os.path.join(test_dir, 'cyclone')
test_dir_e = os.path.join(test_dir, 'earthquake')
test_dir_f = os.path.join(test_dir, 'flood')
test_dir_w = os.path.join(test_dir, 'wildfire')


copy_images = False #False if images are already copied

if copy_images:

    # TRAIN

    names = ['{}.jpg'.format(i) for i in range(400)]
    for name in names:
        source = os.path.join(cyclone_dir, name)
        destination = os.path.join(train_dir_c, name)
        shutil.copyfile(source, destination)
    names = ['{}.jpg'.format(i) for i in range(400)]
    for name in names:
        source = os.path.join(earthquake_dir, name)
        destination = os.path.join(train_dir_e, name)
        shutil.copyfile(source, destination)
    names = ['{}.jpg'.format(i) for i in range(400)]
    for name in names:
        source = os.path.join(flood_dir, name)
        destination = os.path.join(train_dir_f, name)
        shutil.copyfile(source, destination)
    names = ['{}.jpg'.format(i) for i in range(400)]
    for name in names:
        source = os.path.join(wildfire_dir, name)
        destination = os.path.join(train_dir_w, name)
        shutil.copyfile(source, destination)

    # VALIDATION

    names = ['{}.jpg'.format(i) for i in range(400, 500)]
    for name in names:
        source = os.path.join(cyclone_dir, name)
        destination = os.path.join(validation_dir_c, name)
        shutil.copyfile(source, destination)
    names = ['{}.jpg'.format(i) for i in range(400, 500)]
    for name in names:
        source = os.path.join(earthquake_dir, name)
        destination = os.path.join(validation_dir_e, name)
        shutil.copyfile(source, destination)
    names = ['{}.jpg'.format(i) for i in range(400, 500)]
    for name in names:
        source = os.path.join(flood_dir, name)
        destination = os.path.join(validation_dir_f, name)
        shutil.copyfile(source, destination)
    names = ['{}.jpg'.format(i) for i in range(400, 500)]
    for name in names:
        source = os.path.join(wildfire_dir, name)
        destination = os.path.join(validation_dir_w, name)
        shutil.copyfile(source, destination)

    # TEST

    names = ['{}.jpg'.format(i) for i in range(500, 600)]
    for name in names:
        source = os.path.join(cyclone_dir, name)
        destination = os.path.join(test_dir_c, name)
        shutil.copyfile(source, destination)
    names = ['{}.jpg'.format(i) for i in range(500, 600)]
    for name in names:
        source = os.path.join(earthquake_dir, name)
        destination = os.path.join(test_dir_e, name)
        shutil.copyfile(source, destination)
    names = ['{}.jpg'.format(i) for i in range(500, 600)]
    for name in names:
        source = os.path.join(flood_dir, name)
        destination = os.path.join(test_dir_f, name)
        shutil.copyfile(source, destination)
    names = ['{}.jpg'.format(i) for i in range(500, 600)]
    for name in names:
        source = os.path.join(wildfire_dir, name)
        destination = os.path.join(test_dir_w, name)
        shutil.copyfile(source, destination)
